Index: sysdeps/posix/getaddrinfo.c
===================================================================
RCS file: /cvs/glibc/libc/sysdeps/posix/getaddrinfo.c,v
retrieving revision 1.45
diff -u -p -r1.45 getaddrinfo.c
--- sysdeps/posix/getaddrinfo.c	17 Dec 2002 01:14:48 -0000	1.45
+++ sysdeps/posix/getaddrinfo.c	12 Sep 2003 20:26:48 -0000
@@ -51,6 +51,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBI
 #include <sys/utsname.h>
 #include <net/if.h>
 #include <nsswitch.h>
+#include <idna.h>
 
 #define GAIH_OKIFUNSPEC 0x0100
 #define GAIH_EAI        ~(GAIH_OKIFUNSPEC)
@@ -486,6 +487,15 @@ gaih_inet (const char *name, const struc
       at->scopeid = 0;
       at->next = NULL;
 
+      if (req->ai_flags & AI_IDN)
+	{
+	  char *p;
+	  rc = idna_to_ascii_lz (name, &p, 0);
+	  if (rc != IDNA_SUCCESS)
+	    return -EAI_IDN_ENCODE;
+	  name = p; /* XXX memory leak */
+	}
+
       if (inet_pton (AF_INET, name, at->addr) > 0)
 	{
 	  if (req->ai_family == AF_UNSPEC || req->ai_family == AF_INET)
@@ -844,10 +854,10 @@ getaddrinfo (const char *name, const cha
   if (hints == NULL)
     hints = &default_hints;
 
-  if (hints->ai_flags & ~(AI_PASSIVE|AI_CANONNAME|AI_NUMERICHOST))
+  if (hints->ai_flags & ~(AI_PASSIVE|AI_CANONNAME|AI_NUMERICHOST|AI_IDN))
     return EAI_BADFLAGS;
 
-  if ((hints->ai_flags & AI_CANONNAME) && name == NULL)
+  if ((hints->ai_flags & (AI_CANONNAME|AI_IDN)) && name == NULL)
     return EAI_BADFLAGS;
 
   if (service && service[0])
Index: resolv/netdb.h
===================================================================
RCS file: /cvs/glibc/libc/resolv/netdb.h,v
retrieving revision 1.40
diff -u -p -r1.40 netdb.h
--- resolv/netdb.h	8 Aug 2002 21:17:03 -0000	1.40
+++ resolv/netdb.h	12 Sep 2003 20:26:48 -0000
@@ -424,6 +424,11 @@ struct gaicb
 # define AI_PASSIVE	0x0001	/* Socket address is intended for `bind'.  */
 # define AI_CANONNAME	0x0002	/* Request for canonical name.  */
 # define AI_NUMERICHOST	0x0004	/* Don't use name resolution.  */
+# ifdef __USE_GNU
+#  define AI_IDN	0x0008	/* IDN encode input (assuming it is encoded
+                                   in the current locale's character set)
+				   before looking it up. */
+# endif
 
 /* Error values for `getaddrinfo' function.  */
 # define EAI_BADFLAGS	  -1	/* Invalid value for `ai_flags' field.  */
@@ -443,6 +448,7 @@ struct gaicb
 #  define EAI_NOTCANCELED -102	/* Request not canceled.  */
 #  define EAI_ALLDONE	  -103	/* All requests done.  */
 #  define EAI_INTR	  -104	/* Interrupted by a signal.  */
+#  define EAI_IDN_ENCODE  -105	/* IDN encoding of input failed.  */
 # endif
 
 # define NI_MAXHOST      1025
