--- stdint.in.h.orig	2009-01-05 15:02:18.000000000 +0100
+++ stdint.in.h	2009-01-05 15:03:09.000000000 +0100
@@ -478,7 +478,8 @@
    includes <wchar.h> -> <stdio.h> -> <getopt.h> -> <stdlib.h>, and the latter includes
    <stdint.h> and assumes its types are already defined.  */
 #if ! (defined WCHAR_MIN && defined WCHAR_MAX)
-# include <wchar.h>
+/*  We don't need WCHAR_* in libidn, so to avoid problems with
+    missing wchar.h, don't include wchar.h here.  */
 #endif
 #undef WCHAR_MIN
 #undef WCHAR_MAX
